src/features/loading.ts

```diff
@@ -0,0 +1,402 @@
+      this[$posterContainerElement].addEventListener(
+          'click', this[$clickHandler]);
+      this[$posterContainerElement].addEventListener(
+          'keydown', this[$keydownHandler]);
+      this[$progressTracker].addEventListener(
+          'progress', this[$progressHandler]);
+
+      loadingStatusAnnouncer.registerInstance(this);
+    }
+
+    disconnectedCallback() {
+      super.disconnectedCallback();
+
+      this[$posterContainerElement].removeEventListener(
+          'click', this[$clickHandler]);
+      this[$posterContainerElement].removeEventListener(
+          'keydown', this[$keydownHandler]);
+      this[$progressTracker].removeEventListener(
+          'progress', this[$progressHandler]);
+
+      loadingStatusAnnouncer.unregisterInstance(this)
+    }
+
+    async updated(changedProperties: Map<string, any>) {
+      super.updated(changedProperties);
+
+      if (changedProperties.has('poster') && this.poster != null) {
+        this[$defaultPosterElement].style.backgroundImage =
+            `url(${this.poster})`;
+      }
+
+      if (changedProperties.has('src')) {
+        this[$posterDismissalSource] = null;
+        this[$preloadAttempted] = false;
+        this[$lastReportedProgress] = 0;
+        this[$sourceUpdated] = false;
+      }
+
+      if (changedProperties.has('alt')) {
+        this[$defaultPosterElement].setAttribute(
+            'aria-label',
+            `${this[$ariaLabel]}. ${this[$ariaLabelCallToAction]}`);
+      }
+
+      this[$updateLoadingAndVisibility]()
+    }
+
+    [$onClick]() {
+      this[$posterDismissalSource] = PosterDismissalSource.INTERACTION;
+      this.requestUpdate();
+    }
+
+    [$onKeydown](event: KeyboardEvent) {
+      switch (event.keyCode) {
+        // NOTE(cdata): Links and buttons can typically be activated with
+        // both spacebar and enter to produce a synthetic click action
+        case SPACE_KEY:
+        case ENTER_KEY:
+          this[$posterDismissalSource] = PosterDismissalSource.INTERACTION;
+          break;
+        default:
+          break;
+      }
+
+      this.requestUpdate();
+    }
+
+    [$onProgress](event: Event) {
+      const progress = (event as any).detail.totalProgress;
+
+      this.requestUpdate();
+
+      if (progress === 1.0) {
+        this[$updateProgressBar].flush();
+      }
+
+      this[$updateProgressBar](progress);
+
+      this.dispatchEvent(
+          new CustomEvent('progress', {detail: {totalProgress: progress}}));
+
+      this[$lastReportedProgress] =
+          Math.max(progress, this[$lastReportedProgress]);
+    }
+
+    get[$modelIsReadyForReveal](): boolean {
+      const {src} = this;
+
+      return !!src && CachingGLTFLoader.hasFinishedLoading(src) &&
+          this[$lastReportedProgress] === 1.0 && this[$shouldRevealModel]
+    }
+
+    get[$shouldRevealModel](): boolean {
+      return this.reveal === RevealStrategy.AUTO ||
+          this.reveal === this[$posterDismissalSource];
```

Comment:
Would the second clause be clearer as `PosterDismissalSource.INTERACTION === this[$posterDismissalSource]` or `!!this[$posterDismissalSource]`?

Reply:
I suppose my thought (particularly with the second) is that we probably don't care what the dismissal source is, only if we have one?

Reply:
You're right, at this stage we only care that a dismissal source is set. As written, it's only helpful if we had N+1 possible sources.